### DISCUSSION TITLE
Fibers on Truffle - need some pointers

### DISCUSSION THREAD

**Discussion by marad** [oracle/truffleruby#2747] [Q&A] [answered]
Hi guys! Great work with this Ruby implementation!

I've stumbled upon [this post](https://medium.com/graalvm/bringing-fibers-to-truffleruby-1b5d2e258953) talking about fibers in TruffleRuby. 

I'm creating a small language as an experiment and looking forward to implementing algebraic effects there. Having ability to suspend execution would really help here. Can you give me some pointers on how to achieve that? 

I'd very much apreciate any help since, I'm kindof stuck with it right now.

**Comment by nirvdrum** [comment]
TruffleRuby has support for the `Fiber` API. Project Loom, mentioned in the linked post, made its way into Java 19 and that should improve TruffleRuby's Fiber performance. Do you have code that doesn't work with TruffleRuby today?

**Comment by marad** [comment]
Oh no, I guess I haven't been really clear. 

I'm implementing a language based on Truffle and wanted to leverage the continuations to implement algebraic effects. Problem I have is that I don't really know how to achieve this using Truffle.

With TruffleRuby supporting Fibers this is probably achievable even without Project Loom, but I'm not sure how to do this. Do you have any hints?

**Comment by nirvdrum** [comment]
Ahh. The best place to get started is to look at our [FiberNodes](https://github.com/oracle/truffleruby/blob/master/src/main/java/org/truffleruby/core/fiber/FiberNodes.java). You'll be able to navigate to everything of note from there. If you have further questions, it's probably best to hop on the [GraalVM Slack](https://www.graalvm.org/community/) workspace and join the `#truffleruby` or `#truffle` channels. We generally try to keep the issue tracker for bug reports.

**Reply by marad** [reply]
Thanks for the link to FiberNodes that's going to be very informative for me :)

**Comment by eregon** [comment]
Let's convert this to a discussion :)

**Comment by eregon** [comment, marked as answer]
So currently Fibers (= Ruby's stackful coroutines) are implemented using Threads and queues in TruffleRuby, to ensure only one execute at a time (per Ruby Thread).
I'm adding an option to use VirtualThread for Fibers on JDK 19+, but Truffle does not support them well yet, only in the DefaultTruffleRuntime currently.

Regarding using them in Truffle, it's not different than using VirtualThread directly from regular Java, so I think you're mostly asking "what's the API of Loom" and that is Loom's API and that's mostly VirtualThread (e.g. docs https://download.java.net/java/early_access/loom/docs/api/java.base/java/lang/Thread.html).
There is also a lower-level Continuation (https://github.com/openjdk/loom/blob/fibers/src/java.base/share/classes/java/lang/VirtualThread.java#L80) but IIRC that is not public API.

**Reply by marad** [reply]
Thanks, I think this is exactly what I was looking for!
